libs/tiled-bitmap/src/cmyklayeroperations.cc

```diff
@@ -0,0 +1,331 @@
+}
+
+void OperationsCMYK::reduce(Tile::Ptr target, const ConstTile::Ptr source, int top_left_x, int top_left_y)
+{
+  // Reducing by a factor 8
+  int sourceStride = this->bps * source->width / 2; // stride in bytes
+  const byte* sourceBase = source->data.get();
+
+  int targetStride = this->bps * target->width / 2; // stride in bytes
+  byte* targetBase = target->data.get() +
+      (target->height * top_left_y + top_left_x) * targetStride / 8;
+  
+  if (this->bps == 8)
+  {
+    for (int y = 0; y < source->height / 8; y++)
+    {
+      byte* targetPtr = targetBase;
+
+      for (int x = 0; x < source->width / 8; x++)
+      {
+        // We want to store the average colour of the 8*8 pixel image
+        // with (x, y) as its top-left corner into targetPtr.
+        const byte* base = sourceBase + 8 * 4 * x; // start of the row
+        const byte* end = base + 8 * sourceStride; // end of the row
+
+        int sum_c = 0;
+        int sum_m = 0;
+        int sum_y = 0;
+        int sum_k = 0;
+        for (const byte* row = base; row < end; row += sourceStride)
+        {
+          for (size_t current = 0; current < 8 * 4; current += 4)
+          {
+            sum_c += row[current    ];
+            sum_m += row[current + 1];
+            sum_y += row[current + 2];
+            sum_k += row[current + 3];
+          }
+        }
+
+        targetPtr[0] = static_cast<byte>(sum_c / 64);
+        targetPtr[1] = static_cast<byte>(sum_m / 64);
+        targetPtr[2] = static_cast<byte>(sum_y / 64);
+        targetPtr[3] = static_cast<byte>(sum_k / 64);
+
+        targetPtr += 4;
+      }
+
+      targetBase += targetStride;
+      sourceBase += sourceStride * 8;
+    }
+  }
+  else if (this->bps == 4)
+  {
+    for (int y = 0; y < source->height / 8; y++)
+    {
+      byte* targetPtr = targetBase;
+
+      for (int x = 0; x < source->width / 8; x++)
+      {
+        // We want to store the average colour of the 8*8 pixel image
+        // with (x, y) as its top-left corner into targetPtr.
+        const byte* base = sourceBase + 4 * 4 * x; // start of the row
+        const byte* end = base + 8 * sourceStride; // end of the row
+
+        int sum_c = 0;
+        int sum_m = 0;
+        int sum_y = 0;
+        int sum_k = 0;
+        for (const byte* row = base; row < end; row += sourceStride)
+        {
+          for (size_t current = 0; current < 8 * 2; current += 2)
+          {
+            sum_c += row[current    ] >> 4;
+            sum_m += row[current    ] & 15;
+            sum_y += row[current + 1] >> 4;
+            sum_k += row[current + 1] & 15;
+          }
+        }
+
+        targetPtr[0] = static_cast<byte>( sum_c == 15 * 64 ? 0xF0 : (sum_c / 60) << 4)
+                     | static_cast<byte>((sum_m == 15 * 64 ? 0x0F : sum_m / 60));
+        targetPtr[1] = static_cast<byte>( sum_y == 15 * 64 ? 0xF0 : (sum_y / 60) << 4)
+                     | static_cast<byte>((sum_k == 15 * 64 ? 0x0F : sum_k / 60));
```

Comment:
- Why are you dividing those sums by 60?
- Why the special case when the sum equals 15*64?

Reply:
Dividing the sum by 64 (as many of the existing Operation classes do) doesn't quite divide the sum into even groups. Let me give an example for 2bps. Note that the maximum value for the sum is 3 * 64 = 192. For 4bps, the maximum value is 15 * 64.

Division by 64:
sum    0 ... 63 64 ... 127 128 ... 191 192
result 0 ... 0   1 ... 1     2 ... 2   3

"Fairer":
sum    0 ... 47 48 ... 95 96 ... 143 144 ... 191 192
result 0 ... 0   1 ... 1   2 ... 2     3 ... 3   [4 -> 3]

This alternative implementation has made the groups of values that map to a specific value equally big (except for the largest result value). The difference between these alternatives is smaller as the number of bits per sample grows. In this 2bps case, the constants would be 48 and 192. So the constants 60 and 15 * 64 (for 4bps) come from this alternative approach.

Reply:
Ouch
1. That approach could do with a comment. Or better yet, a properly named function that implements this
2. Still confused. 144/60 is 2.4, which becomes 2, not 3 as you suggest in the table
3. To arrive at a sum of 144, you need 48 2's and only 16 3's. That should really become a 2 not a 3.

Summarizing: I'm not opposed to rounding, but I'd like to see a more conventional algorithm. 32 2's and 32 3's surely should become a 3, in the interest of "fairness", but everything lighter than that should be a 2 or less.

Reply:
My example was for 2bps, whereas this code is for 4bps, so the constants would be slightly different. Then the math does work :wink:.

Regardless, this construct is no longer needed when higher layers always use 8bps (as they do since commit 27e92a12244c6fb6408e07e0f78329e48c153ada. As such, I think we can resolve this conversation.